### ISSUE TITLE
[3.3.5] Onyxia spells failing, Eruption missing, Deep Breath not casting

### ISSUE THREAD

**Issue by wetbrownsauce** [TrinityCore/TrinityCore#22824]
Hi, sorry if this report is unwanted, I'm trying to figure all this stuff out still. 

**Description:**
Onyxia has many bugs during the encounter.

**Current behaviour:**
Phase 1:
-Onyxia's Flame Breath and Wing Buffet suffer push back while she is casting them if a player is damaging her.
-Flame Breath and Wing Buffet will sometimes be interrupted by her own auto-attacks.

Phase 2:
-Onyxia will sometimes bug out when using her Deep Breath. The announcement of the ability will still happen on screen, but she will not cast it and then completely bugs out her movement, never flying to a new location until phase 3 when she lands. 
[Gif of the above happening](https://giant.gfycat.com/EasygoingWeeklyArawana.webm)
I'm not too familiar with trinitycore's code, but this could perhaps be a result of the following:

```
void SetNextRandomPoint()
        {
            uint8 MaxCount = sizeof(MoveData) / sizeof(OnyxMove);

            uint8 iTemp = urand(0, MaxCount - 1);

            if (iTemp >= MovePoint)
                ++iTemp;

            MovePoint = iTemp;
        }
```

Unless I'm misunderstanding this, it seems that if iTemp was randomized to 7, and Onyxia's previous MovePoint ended on 7 via `MovePoint = PointData->LocIdEnd;`, then iTemp would be incremented to 8 resulting in a MovePoint of 8 which then returns null later in the code when called by `GetMoveData()`

Please ignore that rambling if I'm entirely wrong, I'm pretty new to this. 

-When landing after phase 2, Onyxia floats until kited around a bit.
[Gif of the above happening](https://giant.gfycat.com/WarlikeDistantBarracuda.webm)

Phase 3:
-The Eruption of lava from cracks in the floor does no damage. In the combat log it repeatedly says "Lava Fissure's Eruption misses Onyxia" and once Onyxia is dead, it begins saying this about the player. 
[Image of combat log in phase 3](https://i.imgur.com/qJMNR5Y.png)
-Flame Breath, Wing Buffet, AND Bellowing Roar all suffer the same push back and interrupts during this phase.
[Gif of the above happening, this occurs in both phase 1 & 3](https://giant.gfycat.com/KlutzyScentedBoa.webm) 
-Whelps do not spawn in phase 3 at all, even though they should. (I think this is already known though).

**Branch(es):**
3.3.5

**TC rev. hash/commit:** 
[352c4a4]

**Comment by D0d0** [comment]
@Killyana could you please refer me to and gameobject rewrite? Because there is a similar issue with Heigans erruption (it does not do damage) and I think, it has something to do with that rewrite.

**Comment by Killyana** [comment]
It's not really the cast from the game object https://github.com/TrinityCore/TrinityCore/pull/21395 but the way how the gameobject is created, if you gob add temp the gob it will probably cast the spell on the player correctly, but for spawned gobs it will not.
The issue exist since probably ever but was hidden by the cast hack that was used for gameobjects.

**Comment by Jildor** [comment]
Problems with pushback cast spells should be fixed by: 37f69825c445f3c973172792e1abc0bdc9f02dde
Problems with Movements should be by a movement rewrite commits maybe (generic problem?)
Problems with gameobjects should be by a gameobject rewrite, generic problem

Maybe open a separated issues if movement and gameobject problems are generics it would be better?

How @Killyana says,  Onyxia was working correctly before these rewrites :(

**Comment by Jildor** [comment]
This do lava fissure cast spell:
```diff
diff --git a/src/server/scripts/Kalimdor/OnyxiasLair/instance_onyxias_lair.cpp b/src/server/scripts/Kalimdor/OnyxiasLair/instance_onyxias_lair.cpp
index 1910684c12..4fa64d1e60 100644
--- a/src/server/scripts/Kalimdor/OnyxiasLair/instance_onyxias_lair.cpp
+++ b/src/server/scripts/Kalimdor/OnyxiasLair/instance_onyxias_lair.cpp
@@ -121,7 +121,12 @@ class instance_onyxias_lair : public InstanceMapScript
                 //THIS GOB IS A TRAP - What shall i do? =(
                 //Cast it spell? Copyed Heigan method
                 floorEruption->SendCustomAnim(floorEruption->GetGoAnimProgress());
-                floorEruption->CastSpell(nullptr, Difficulty(instance->GetSpawnMode()) == RAID_DIFFICULTY_10MAN_NORMAL ? 17731 : 69294); //pFloorEruption->GetGOInfo()->trap.spellId
+                CastSpellExtraArgs args;
+                args.OriginalCaster = onyxiaGUID;
+                floorEruption->CastSpell(floorEruption, floorEruption->GetGOInfo()->trap.spellId, args);
 
                 //Get all immediatly nearby floors
                 std::list<GameObject*> nearFloorList;
```

But for some reason, the spell hit onyxia too, I don't know why

**Comment by Killyana** [comment]
Maybe wrong gob faction?

**Comment by Jildor** [comment]
I think gob faction is 14 (Monster), those values are not sniffed?
or can be change? I don't know what faction should have

**Comment by Jinnaix** [comment]
The correct faction would be 80. Seems like if creatures spawn gameobjects, they dont inherit their faction correctly?

**Comment by Killyana** [comment]
Onyxia use the faction 103

**Comment by Jildor** [comment]
But traps are spawned by DB @Jinnaix

**Comment by Jinnaix** [comment]
Funny, i looked up the faction over aowow, its 80 there. But you are right ofc Kilyana.
Well, gambeoject_template_addon it is then. 
The script itself does a pretty lazy way of getting the required gameobjects by just targetting the display id 4392.  That would be 74 Gob`s in total.

**Comment by Killyana** [comment]
faction 14 is correct:
``` 
[8] UpdateType: CreateObject1
[8] GUID: Full: 0xF112B2B8000010AB Type: GameObject Entry: 176824 Low: 4267
[8] Object Type: 5 (GameObject)
[8] Living: False
[8] Has Attacking Target: False
[8] Has Vehicle Data: False
[8] Has Stationary Position: True
[8] Has GameObject Position: False
[8] Has AnimKits: False
[8] Has GameObject Rotation: True
[8] GO Rotation: X: 0 Y: 0 Z: -0.7660437 W: 0.6427886
[8] O: 4.537859
[8] Stationary Position: X: 31.16722 Y: -218.9902 Z: -83.82716
[8] OBJECT_FIELD_GUID: Full: 0xF112B2B8000010AB Type: GameObject Entry: 176824 Low: 4267
[8] OBJECT_FIELD_TYPE: 33
[8] OBJECT_FIELD_ENTRY: 176824
[8] OBJECT_FIELD_SCALE_X: 1
[8] GAMEOBJECT_DISPLAYID: 4392/6.154503E-42
[8] GAMEOBJECT_PARENTROTATION + 2: 1065353216/1
[8] GAMEOBJECT_PARENTROTATION + 3: 3007036718/-4.371139E-08
[8] GAMEOBJECT_DYNAMIC: 4294901760/NaN
[8] GAMEOBJECT_FACTION: 14
[8] GAMEOBJECT_LEVEL: 80/1.121039E-43
[8] GAMEOBJECT_BYTES_1: 1/6/0/255

**Comment by Jinnaix** [comment]
So we miss a lot of general checks for spell -casted by gameobjects- targetting.

**Comment by Killyana** [comment]
Spell: 17731 (heroic: 69294)  is used by GameObject Entry: 176824
Probably just a condition on effect 0 to hit only players will fix the issue, but still we have an other issue that spams the log: https://github.com/TrinityCore/TrinityCore/issues/11877
``` 
Effect 0: Id 2 (SPELL_EFFECT_SCHOOL_DAMAGE)
BasePoints = 5183 to 5816
Targets (18, 16) (TARGET_DEST_CASTER, TARGET_UNIT_DEST_AREA_ENEMY)
Radius (Id 8) 5,00

Effect 1: Id 86 (SPELL_EFFECT_ACTIVATE_OBJECT)
BasePoints = 1 Bonus = 1,00
Targets (18, 52) (TARGET_DEST_CASTER, TARGET_GAMEOBJECT_DEST_AREA)
EffectMiscValueA = 5
Radius (Id 17) 13,00

**Comment by Jildor** [comment]
@Killyana condition for that, can be affect player pets too? because player pets should be affected by eruption spell
We can do a spell script for eruption and filtertargets, but I don't know if can be hacky

**Comment by Killyana** [comment]
Actually no, something is wrong with gobs checking factions:
Try to find 2 aggressive npcs:
.mod fac 14    the first
.mod fac 103  the second
Select the first and .cast self 17731
notice you get hit but not the other npc
do the same with the second npc and you only get hit
If you cast the spell in your self you will hit both pcs

So the issue must be fixed in core side for gobs faction "check"

**Comment by Jinnaix** [comment]
Proposal: 
New Issue for the basic core problem (couldnt find one covering this).
Temporary adjustment with gob_temp_addon with reference in the new issue to have it at least working till core support is finished.

**Comment by Jildor** [comment]
Maybe we should include in that check gameobjects too:
https://github.com/TrinityCore/TrinityCore/blob/0468c70dfe91794ad272594323dd7feb611d0a93/src/server/game/Entities/Object/Object.cpp#L2900-L2902

something like:
```diff
@@ bool WorldObject::IsValidAttackTarget(WorldObject const* target, SpellInfo const
     }
 
     // CvC case - can attack each other only when one of them is hostile
-    if (unit && !HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PLAYER_CONTROLLED) && target->ToUnit() && !target->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PLAYER_CONTROLLED))
+    if (((unit && !HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PLAYER_CONTROLLED)) || ToGameObject()) && target->ToUnit() && !target->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_PLAYER_CONTROLLED))
         return IsHostileTo(target) || target->IsHostileTo(this);
 
     // PvP, PvC, CvP case
```

but need more research, @joshwhedon what do you think?

**Comment by Jildor** [comment]
Traps shouldn't hit Onyxia now, what problem remain on this issue?

**Comment by Killyana** [comment]
This issue got fixed long time ago.
